connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java

```diff
@@ -0,0 +1,524 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.header;
+
+import org.apache.kafka.common.utils.AbstractIterator;
+import org.apache.kafka.connect.data.Date;
+import org.apache.kafka.connect.data.Decimal;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.data.Time;
+import org.apache.kafka.connect.data.Timestamp;
+import org.apache.kafka.connect.errors.DataException;
+
+import java.math.BigDecimal;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A basic {@link Headers} implementation.
+ */
+public class ConnectHeaders implements Headers {
+
+    private static final int EMPTY_HASH = Objects.hash(new LinkedList<>());
+
+    // Shared, immutable instance with a null schema and null value
+    private static final SchemaAndValue NULL_SCHEMA_AND_VALUE = new SchemaAndValue(null, null);
+
+    /**
+     * An immutable and therefore sharable empty iterator.
+     */
+    private static final Iterator<Header> EMPTY_ITERATOR = new Iterator<Header>() {
+        @Override
+        public boolean hasNext() {
+            return false;
+        }
+
+        @Override
+        public Header next() {
+            throw new NoSuchElementException();
+        }
+
+        @Override
+        public void remove() {
+            throw new IllegalStateException();
+        }
+    };
+
+
+    // This field is set lazily, but once set to a list it is never set back to null
+    private volatile LinkedList<Header> headers;
+
+    public ConnectHeaders() {
+    }
+
+    public ConnectHeaders(Iterable<Header> original) {
+        if (original != null) {
+            if (original instanceof ConnectHeaders) {
+                ConnectHeaders headers = (ConnectHeaders) original;
+                if (!headers.isEmpty()) {
+                    this.headers = new LinkedList<>(((ConnectHeaders) original).headers);
+                }
+            } else {
+                this.headers = new LinkedList<>();
+                for (Header header : original) {
+                    this.headers.add(header);
+                }
+            }
+        }
+    }
+
+    @Override
+    public int size() {
+        return this.headers == null ? 0 : headers.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return this.headers == null ? true : headers.isEmpty();
```

Comment:
the norm (ideally) for AK is to only use `this.` as needed to handle masking. i'm sure there are exceptions, but almost universally the use case is for constructors to avoid awkward parameter naming in order to avoid conflicting with a convenient member variable name.

Reply:
Well, at least we should be consistent within a compilation unit, so I'll fix that.